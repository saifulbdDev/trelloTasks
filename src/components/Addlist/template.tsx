/* eslint-disable eqeqeq */
/* eslint-disable jsx-a11y/alt-text */
/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { useState, ChangeEvent } from "react";
import "./_styles.scss";
import closeIcon from "../../styles/images/kindpng_2148901.png";
interface ICardProps {
  boardStore: any;
  createColumn: any;
}
interface IColumnState {
  boardStore: any;
}

export const Addlist: React.FC<ICardProps & IColumnState> = (props) => {
  const [addList, setList] = React.useState<boolean>(false);
  const [listName, setNewname] = useState("");
  const [nameError, setError] = useState("");
  const titile = Object.keys(props.boardStore).length
    ? "Add another list"
    : "Add a list";

  const nameChange = (e: ChangeEvent<HTMLInputElement>) => {
    setNewname(e.target.value);
    setError("");
  };

  const handleOnAdd = () => {
    setList(true);
  };
  const handleOnClose = () => {
    setList(false);
  };
  const handleAdd = () => {
    if (listName == "") {
     
    } else {
      if (props.boardStore.hasOwnProperty('column_'+listName)) {
        setError("This list name already exist.");
      } else {
         
        props.createColumn('column_'+listName);
        setNewname("");
      }
    }
  };

  const addListFrom = () => {
    return (
      <div className={`addlist-from ` + (nameError ? "error-from" : "")}>
        <input
          type="text"
          className="addlist-input"
          placeholder="Enter list titleâ€¦"
          onChange={nameChange}
          value={listName}
        />
        <div className="error-text">{nameError}</div>
        <div className="addlist-from-btn-group">
          <button className="addlist-submit" onClick={handleAdd}>
            Add List
          </button>
          <a href="#"  className="addlist-close" onClick={handleOnClose}>
            <img width="20" src={closeIcon} alt="close icon" />
          </a>
        </div>
      </div>
    );
  };
  const addListButton = () => {
    return (
      <span className="addlist-btn" onClick={handleOnAdd}>
        <i className="fas fa-plus"></i> {titile}
      </span>
    );
  };

  const content = addList ? addListFrom() : addListButton();

  return (
    <div
      className={`addlist-container list ` + (addList ? "" : "addlist-active")}
    >
      {content}
    </div>
  );
};
