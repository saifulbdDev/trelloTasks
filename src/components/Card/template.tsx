import React from "react";
import "./styles.scss";

interface ICardProps {
  category: string;
  categoryTasks?: any;
  createCard?: any;
  removeCard?: any;
  updateCard?: any;
  editable?: boolean;
  text?: string;
  value?: string;
  id: number;
}

interface ICardState {
  id?: number;
  text?: string;
  value?: string;
  editable?: boolean;
}

export const Card: React.FC<ICardProps> = (props) => {
  const [state, setState] = React.useState<ICardState>(props);
  const [isDragged, setIsDragged] = React.useState<boolean>(false);

  React.useEffect(() => {
    setState({ editable: false, value: "", text: "", id: 0 });
  }, [props.category]);

  const handleOnEdit = () => {
    setState({ editable: true });
  };

  const handleOnRemove = () => {
    props.removeCard(props.id);
  };

  const handleOnBlur = (event: { target: { value: string; }; }) => {
    const cleanedValue = event.target.value.trim();
    if (cleanedValue !== "") {
      setState({ editable: false, value: event.target.value });
      props.updateCard({ id: props.id, text: state.text });
    }
  };

  const handleChange = (event: any) => {
    setState({ value: event.target.value });
  };

  const handleOnDragStart = (ev: any, id: number) => {
    ev.dataTransfer.setData("id", id);
    ev.dataTransfer.setData("sourceCategory", props.category);
    setIsDragged(true);
  };

  const handleOnDragStop = () => {
    setIsDragged(false);
  }

  const getInputField = () => {
    return (
      <textarea
        value={state.value}
        onChange={handleChange}
        onBlur={handleOnBlur}
      />
    );
  }

  const content = state.editable
    ? getInputField()
    : props.text;

  return (
    <div
      className={`card ${props.category}` + (isDragged ? "isDragged" : "")}
      onDragStart={(e) => handleOnDragStart(e, props.id)}
      onDragEnd={handleOnDragStop}
      draggable
    >
      <button className="button-remove" onClick={handleOnRemove}>âœ–</button>
      <div className="card-body" onClick={handleOnEdit}>
        {content}
      </div>
    </div>
  );

}