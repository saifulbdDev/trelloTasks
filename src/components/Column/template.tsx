
import React, { useState, ChangeEvent, useEffect, useRef } from "react";
import { Outside } from "../../common/outside";
import { v4 as uuidv4 } from "uuid";
import closeIcon from "../../styles/images/kindpng_2148901.png";
import { Card } from "../Card/";
import "./_styles.scss";

interface IColumnProps {
  category: any;
  catkey: any;
  categoryTasks: any;
  createCard: any;
  updateCard: (card: any, destinationCategory: string) => void;
  updateColumn: (upColumn: string, oldColumn: string) => void;
  dateleColumn: (catId: string) => void;
  moveCard: (
    cardId: string,
    sourceCategory: string,
    destinationCategory: string
  ) => void;
  removeCardFromColumn: (id: number, category: string) => void;
}

interface IColumnState {
  category: string;
  categoryTasks: string[];
}

export const Column: React.FC<IColumnProps & IColumnState> = (props) => {
  const [state, setState] = useState<IColumnState>(props);
  const wrapperRef = useRef(null);
  const [newtext, setNewtext] = useState("");
  const categoryTasks = props.categoryTasks;
  const category = props.category ? props.category : "";
  const [addCard, setAddcard] = React.useState<boolean>(false);
  const [updateTitle, setTitle] = useState(category.title);
  const [titleEditble, setTitleEditble] = React.useState<boolean>(false);
  const textChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
    setNewtext(e.target.value);
  };
  const titleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setTitle(e.target.value);
  };

  Outside(wrapperRef, () => titileOnSave());

  useEffect(() => {
    setState({
      category: props.category,
      categoryTasks: [],
    });
  }, [props.category]);
  const titileOnUpdate = () => {
    setTitleEditble(true);
  };

  const titileOnSave = () => {
    setTitleEditble(false);

    if (updateTitle !== category.title && updateTitle !== "") {
      console.log("updateTitle", updateTitle);
      props.updateColumn(updateTitle, category.id);
    } else {
      setTitle(category.title);
    }
  };

  const columnOnDelete = () => {
    props.dateleColumn(category.id);
  };
  const handleOnAdd = () => {
    setAddcard(true);
  };
  const handleOnClose = () => {
    setAddcard(false);
  };
  const handleAdd = () => {
    if (newtext) {
      const new_task = { id: uuidv4(), text: newtext };
      const categoryTasks = [new_task, ...props.categoryTasks];

      setState({
        ...state,
        category: state.category,
        categoryTasks: categoryTasks,
      });

      props.createCard(new_task, props.catkey);
      setNewtext("");
    }
  };

  const handleRemove = (cardIdToRemove: number) => {
    const categoryTasks = props.categoryTasks.filter((task: any) => {
      return task.id !== cardIdToRemove;
    });
    setState({ ...state, categoryTasks: categoryTasks });
    props.removeCardFromColumn(cardIdToRemove, props.catkey);
  };

  const handleUpdate = (card: any) => {
    const new_task = card;
    const categoryTasks = [new_task, ...props.categoryTasks];
    setState({ ...state, categoryTasks: categoryTasks });
    props.updateCard(new_task, props.category);
  };

  const handleOnDragOver = (event: any) => {
    event.preventDefault();
  };

  const handleOnDrop = (event: any) => {
    let id = event.dataTransfer.getData("id");
    let sourceCategory = event.dataTransfer.getData("sourceCategory");
    moveTask(id, sourceCategory);
  };

  const moveTask = (taskId: string, sourceCategory: string) => {
    if (sourceCategory !== props.catkey) {
      props.moveCard(taskId, sourceCategory, props.catkey);
    }
  };

  const cards = categoryTasks.map((task: any, key: number) => (
    <Card
      key={key}
      id={task.id}
      text={task.text}
      category={props.catkey}
      removeCard={handleRemove}
      updateCard={handleUpdate}
    />
  ));

  const addListFrom = () => {
    return (
      <div className="add-from ">
        <textarea
          placeholder="Enter a title for this cardâ€¦"
          value={newtext}
          onChange={textChange}
        />

        <div className="cardfrom-btn-group">
          <button className="addcard-submit" onClick={handleAdd}>
            Add card
          </button>
          <a href="#" className="addcard-close" onClick={handleOnClose}>
            <img width="20" src={closeIcon} alt="close icon" />
          </a>
        </div>
      </div>
    );
  };
  const addListButton = () => {
    return (
      <div className="addcard-btn-group">
        <a href="#" className="addcard-btn" onClick={handleOnAdd}>
          <i className="fas fa-plus"></i> Add card
        </a>
        <span className="addcard-icon">
          <i className="fas fa-file-video"></i>
        </span>
      </div>
    );
  };

  const content = addCard ? addListFrom() : addListButton();

  return (
    <div className="list" onDrop={handleOnDrop} onDragOver={handleOnDragOver}>
      <div className="list-header">
        <div className="list-header-contant">        
          <input
            className={titleEditble ? "list-header-field" : "list-header-title"}
            ref={wrapperRef}
            onChange={titleChange}
            onDoubleClick={titileOnUpdate}
            readOnly={!titleEditble}
            onKeyUp={(e) => {
              if (e.key === "Enter") {
                titileOnSave();
              }
            }}
            value={updateTitle}
          />
        </div>
        <div className="list-header-actions">
          <a href="#" onClick={columnOnDelete} className="list-header-action">
            <i className="fas fa-trash-alt"></i>
          </a>
        </div>
      </div>

      <div className="task-container droppable">
        {cards}
        {content}
      </div>
    </div>
  );
};
