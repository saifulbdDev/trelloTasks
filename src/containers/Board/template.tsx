import React from "react";
import "./Board.scss";
import { connect } from "react-redux";
import { fetchData } from "../../store/actions/actions";
import { Column } from "../../components/Column/template";


export interface IAppState {
  boardStore: any;
}

interface IBoardProps {
  fetchData: () => void;
  removeCardFromColumn: (cardId: number, sourceCategory: string) => void;
  moveCard: (cardId: number, sourceCategory: string, destinationCategory: string) => void;
  updateCard: (card: any, destinationCategory: string) => void;
  createCard?: (card: any, destinationCategory: string) => void;
  boardStore: any;
}

interface IBoardState {
  boardStore: any;
}

export const Board: React.FC<IBoardProps & IBoardState> = (props: IBoardProps) => {
  const [, setState] = React.useState<IBoardState>(props);

  const { boardStore, fetchData } = props;

  // const handleAdd = () => {
  //   // const maxIdPerCategory = Math.max(
  //   //   ...props.categoryTasks.map((item: any) => item.id),
  //   //   0
  //   // );
  //   const nextId = maxIdPerCategory + 1;
  //   const cardText = `${"Card" + nextId}`;
  //   const new_task = { id: nextId, text: cardText };
  //   const categoryTasks = [new_task, ...props.categoryTasks];
  //   setState({ ...state, category: state.category, categoryTasks: categoryTasks });
  //   props.createCard(new_task, props.category);
  // };

  React.useEffect(() => {
    setState({
      boardStore: {}
    });
  }, [boardStore]);

  React.useEffect(() => {
    fetchData();
  }, [fetchData]);

  console.log("Board: ", boardStore);
  const boardColumns = (props.boardStore && Object.keys(props.boardStore)) || [];
  const columns = boardColumns.map((category: string, key: number) => {
    return (
      <Column
        key={key}
        category={category}
        removeCardFromColumn={props.removeCardFromColumn}
        moveCard={props.moveCard}
        createCard={props.createCard}
        updateCard={props.updateCard}
        categoryTasks={props.boardStore[category]}
      />
    );
  });

  return <div className="Board board-lists"> {columns} <div><input type="text" placeholder="fnfnfn"/></div> </div>;
}


export default connect(
  (state: IAppState) => {
    let boardStore;
    if (state.boardStore.constructor === Array) {
      boardStore = {};
    } else boardStore = state.boardStore;
    return { boardStore };
  },
  { fetchData }
)(Board);
