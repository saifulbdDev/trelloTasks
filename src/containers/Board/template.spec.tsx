import React from "react";
import { Board }  from "./template";
import { shallow } from "enzyme";

describe("Board", () => {
  const props = {
    category: "abc",
    createCard: jest.fn(),
    fetchData: jest.fn(),
    moveCard: jest.fn(),
    updateCard: jest.fn(),
    removeCardFromColumn: jest.fn(),
    boardStore: {}
  };
  let board = shallow(<Board {...props} />);
  it("renders properly", () => {
    expect(board).toMatchSnapshot();
  });
});
