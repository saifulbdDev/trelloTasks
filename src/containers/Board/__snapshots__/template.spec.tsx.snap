// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders properly 1`] = `ShallowWrapper {}`;
